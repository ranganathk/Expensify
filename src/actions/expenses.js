import uuid from 'uuid';

export const addExpense = ({ description = '', note = '', createdAt = 0, amount = 0, paidTo = '' } = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    note,
    createdAt,
    paidTo,
    amount
  }
});

export const deleteExpense = ({id} = {}) => ({
  type: 'DELETE_EXPENSE',
  expense: {
    id
  }
});

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
