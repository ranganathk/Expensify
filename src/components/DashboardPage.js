import React from 'react';
import ExpensesList from './ExpensesList';

const DashboardPage = () => (
  <div>
    <ExpensesList />
  </div>
);

export default DashboardPage;
