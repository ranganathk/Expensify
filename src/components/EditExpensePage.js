import React from 'react';

const EditExpensePage = () => (
  <div>
    Hi I am in edit expense page.
  </div>
);

export default EditExpensePage;
