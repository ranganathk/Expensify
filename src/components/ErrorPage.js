import React from 'react';
import { Link } from 'react-router-dom';

const ErrorPage = () => (
  <div>
    <p>Sorry!!! The page you are looking for cannot be found.</p>
    <Link to="/">Go to home page</Link>
  </div>
);

export default ErrorPage;
