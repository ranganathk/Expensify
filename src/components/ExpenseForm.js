import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
  state = {
    description: '',
    amount: '',
    paidTo: '',
    note: '',
    createdAt: moment(), 
    calenderFocused: false,
    error: ''
  };

  changeDescription = (ev) => {
    const description = ev.target.value;
    this.setState(() => ({ description }));
  };

  changePayee = (ev) => {
    const paidTo = ev.target.value;
    this.setState(() => ({ paidTo }));
  };

  changeNote = (ev) => {
    const note = ev.target.value;
    this.setState(() => ({ note }));
  };

  changeAmount = (ev) => {
    const amount = ev.target.value;
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  changeDate = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  focusOnDate = ({focused}) => {
    this.setState(() => ({ calenderFocused: focused }));
  };

  addExpense = (ev) => {
    ev.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Kindly add a description and amount.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseInt(this.state.amount),
        paidTo: this.state.paidTo,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf()
      })
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form 
          onSubmit={this.addExpense}
        >
          <input 
            type="text"
            placeholder="description"
            value={this.state.description}
            onChange={this.changeDescription}
            autoFocus
          />
          <input 
            type="text"
            placeholder="amount"
            value={this.state.amount}
            onChange={this.changeAmount}
          />
          <input 
            type="text"
            placeholder="payee"
            value={this.state.paidTo}
            onChange={this.changePayee}
          />
          <SingleDatePicker 
            date={this.state.createdAt}
            onDateChange={this.changeDate} 
            focused={this.state.calenderFocused}
            onFocusChange={this.focusOnDate}
            isOutsideRange={(day) => false}
            numberOfMonths={1}
          />
          <textarea 
            placeholder="Add a note (optional)"
            value={this.state.note}
            onChange={this.changeNote}
          >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  };
};