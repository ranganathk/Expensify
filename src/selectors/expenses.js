import moment from 'moment';

export default (expenses, { text, startDate, endDate, sortBy }) => {
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;

    return startDateMatch && endDateMatch && textMatch;
  }).sort((expense1, expense2) => {
    if (sortBy === 'date') {
      return expense1.createdAt < expense2.createdAt;
    } else if (sortBy === 'amount') {
      return expense1.amount < expense2.amount;
    }
  })
};
