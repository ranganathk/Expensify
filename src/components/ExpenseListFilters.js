import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  };

  changeDates = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };

  focusOnDate = (calenderFocused) => {
    this.setState(() => ({ calenderFocused }))
  }

  render() {
    return (
      <div>
        <input 
          value={this.props.filters.text} 
          onChange={(e) => ( 
            this.props.dispatch(setTextFilter(e.target.value))
          )}
        />
        <select 
          value={this.props.filters.sortBy} 
          onChange={(e) => {
            if (e.target.value === 'amount') {
              return this.props.dispatch(sortByAmount());
            } else {
              return this.props.dispatch(sortByDate());
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker 
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.changeDates} 
          focusedInput={this.state.calenderFocused}
          onFocusChange={this.focusOnDate}
          isOutsideRange={(day) => false}
          numberOfMonths={1}
          showClearDates={true}
        />
      </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
};

export default connect(mapStateToProps)(ExpenseListFilters);
