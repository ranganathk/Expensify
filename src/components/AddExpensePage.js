import React from 'react';
import ExpenseForm from './ExpenseForm';

const AddExpensePage = () => (
  <div>
    <h3>Add an expense in this page.</h3>
    <ExpenseForm />
  </div>
);

export default AddExpensePage;
