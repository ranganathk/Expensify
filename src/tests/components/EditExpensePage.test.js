import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let wrapper, history, editExpense, deleteExpense;

beforeEach(() => {
  editExpense = jest.fn();
  deleteExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<EditExpensePage expense={expenses[2]} history={history} editExpense={editExpense} deleteExpense={deleteExpense} />);
});

test('should render add expense page correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit of edit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
});

test('should handle onClick of delete', () => {
  wrapper.find('button').simulate('click')
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(deleteExpense).toHaveBeenLastCalledWith({id: expenses[2].id});
});