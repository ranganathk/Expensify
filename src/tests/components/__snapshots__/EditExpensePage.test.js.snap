// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add expense page correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 400,
        "createdAt": -432000000,
        "description": "Gas Bill",
        "id": "3",
        "note": "",
        "paidTo": "Bharat Gas",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
